metadata/src/main/java/org/apache/kafka/metadata/authorizer/ClusterMetadataAuthorizer.java

```diff
@@ -0,0 +1,132 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.metadata.authorizer;
+
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.acl.AclBinding;
+import org.apache.kafka.common.acl.AclBindingFilter;
+import org.apache.kafka.common.errors.ApiException;
+import org.apache.kafka.common.errors.UnknownServerException;
+import org.apache.kafka.common.requests.ApiError;
+import org.apache.kafka.server.authorizer.AclCreateResult;
+import org.apache.kafka.server.authorizer.AclDeleteResult;
+import org.apache.kafka.server.authorizer.AuthorizableRequestContext;
+import org.apache.kafka.server.authorizer.Authorizer;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+
+/**
+ * An interface for Authorizers which store state in the __cluster_metadata log.
+ *
+ * These methods must all be thread-safe.
+ */
+public interface ClusterMetadataAuthorizer extends Authorizer {
+    /**
+     * Set the mutator object which should be used for creating and deleting ACLs.
+     */
+    void setAclMutator(AclMutator aclMutator);
+
+    /**
+     * Get the mutator object which should be used for creating and deleting ACLs.
+     *
+     * @throws org.apache.kafka.common.errors.NotControllerException
+     *              If the aclMutator was not set.
+     */
+    AclMutator aclMutatorOrException();
+
+    /**
+     * Load the ACLs in the given map. Anything not in the map will be removed.
+     * The authorizer will also wait for this initial snapshot load to complete when
+     * coming up.
+     */
+    void loadSnapshot(Map<Uuid, StandardAcl> acls);
+
+    /**
+     * Add a new ACL. Any ACL with the same ID will be replaced.
+     */
+    void addAcl(StandardAcl acl);
+
+    /**
+     * Remove the ACL with the given ID.
+     */
+    void removeAcl(Uuid id);
+
+    /**
+     * Returns true if this authorizer contains an ACL equivalent to the specified one.
+     */
+    boolean hasEquivalentAcl(StandardAcl acl);
+
+    default List<? extends CompletionStage<AclCreateResult>> createAcls(
+            AuthorizableRequestContext requestContext,
+            List<AclBinding> aclBindings) {
+        List<CompletableFuture<AclCreateResult>> futures = new ArrayList<>(aclBindings.size());
+        AclMutator aclMutator = aclMutatorOrException();
+        aclBindings.forEach(b -> futures.add(new CompletableFuture<>()));
+        aclMutator.createAcls(aclBindings).whenComplete((results, throwable) -> {
+            if (throwable == null && results.size() != futures.size()) {
```

Comment:
if it has to check `futures`'s `size` every time, what about assign one time and use many times?

```suggestion
            final int futuresSize = futures.size();
            if (throwable == null && results.size() != futuresSize) {
```

Reply:
Finding the size of an ArrayList takes O(1) time, so the suggestion given here does not improve performance.